### PR TITLE
[Sparse] Add matrix slicing python API

### PR DESCRIPTION
**Pull request by xiangyuzhi**
## Description
Define the python interfaces for matrix slicing operators, includes:  rowwise_select(ids), columnwise_select(ids), rowwise_select(start, end), columnwise_select(start, end)

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/sparse/sparse_matrix.py [hunk 1]
```diff
@@ -479,6 +479,128 @@ def is_diag(self):
...
+        >>> A = dglsp.spmatrix(indices, val)
+        >>> row_ids = torch.tensor([0, 1, 4])
+        >>> B = A.rowwise_select(row_ids)
+        >>> B
+        SparseMatrix(indices=tensor([[0, 1, 1, 2],
+                                     [0, 2, 4, 0]]),
+                     values=tensor([0, 1, 2, 5]),
+                     shape=(3, 6), nnz=4)
+        """
+        return SparseMatrix(self.c_sparse_matrix.rowwise_select(ids))
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:510]: Does this exist?
If not, just raise an not implemented error.

File: python/dgl/sparse/sparse_matrix.py [hunk 2]
```diff
@@ -479,6 +479,128 @@ def is_diag(self):
         """Returns whether the sparse matrix is a diagonal matrix."""
         return self.c_sparse_matrix.is_diag()
 
+    def rowwise_select(self, ids: torch.Tensor):
+        """Returns a row-wise sparse matrix selected by ids.
+
+        Parameters
+        ----------
+        ids : tensor.Tensor
+            The indices indicate the rows selected from matrix.
```
Comment (czkkkkkk) [inline python/dgl/sparse/sparse_matrix.py:488]:
```suggestion
            The row ids to select from matrix.
```

File: python/dgl/sparse/sparse_matrix.py [hunk 3]
```diff
@@ -479,6 +479,128 @@ def is_diag(self):
...
+        SparseMatrix
+            The row-wise sparse matrix only contains the rows selected.
+
+        Examples
+        --------
+        >>> indices = torch.tensor([0, 1, 1, 2, 3, 4], [0, 2, 4, 3, 5, 0]])
+        >>> val = torch.tensor([0, 1, 2, 3, 4, 5])
+        >>> A = dglsp.spmatrix(indices, val)
+        >>> row_ids = torch.tensor([0, 1, 4])
+        >>> B = A.rowwise_select(row_ids)
```
Comment (czkkkkkk) [inline python/dgl/sparse/sparse_matrix.py:503]:
```suggestion
        >>> A.rowwise_select(row_ids)
```

And remove the next line.

File: python/dgl/sparse/sparse_matrix.py [hunk 4]
```diff
@@ -479,6 +479,128 @@ def is_diag(self):
...
+        >>> B
+        SparseMatrix(indices=tensor([[0, 1, 1, 2],
```
Comment (czkkkkkk) [inline python/dgl/sparse/sparse_matrix.py:505]: The example looks incorrect. Please rerun the example after your implement this interface.

**Review by czkkkkkk — COMMENTED** [review]
I'd suggest you to try Github copilot to improve your wording and coding efficiency.

**Comment by xiangyuzhi** [comment]
Now we have two API designs, one is row-column-select API, and another is index-range-select API.
row-column-select API:
```
def rowwise_select(index or range) # User decides whether index or range by input.
def columnwise_select(index or range)
```
index-range-select API:
```
def Index_select(dim, index) # User decides whether row or column by 'dim'.
def Range_select(dim, range) 
```
Or even we can combine them all like:
```
def select(dim, index or range) # User decides dim and index/range.
```
@frozenbugs do you have some comments?

**Comment by frozenbugs** [comment]
> Now we have two API designs, one is row-column-select API, and another is index-range-select API. row-column-select API:
> 
> ```
> def rowwise_select(index or range) # User decides whether index or range by input.
> def columnwise_select(index or range)
> ```
> 
> index-range-select API:
> 
> ```
> def Index_select(dim, index) # User decides whether row or column by 'dim'.
> def Range_select(dim, range) 
> ```
> 
> Or even we can combine them all like:
> 
> ```
> def select(dim, index or range) # User decides dim and index/range.
> ```
> 
> @frozenbugs do you have some comments?

The index/range should just be tensor, right? And it is quite common to use dim in dgl Sparse API to indicate row wise and column wise, see (https://docs.dgl.ai/en/1.1.x/api/python/dgl.sparse_v0.html), so def select(dim, Tensor) should be good.

File: python/dgl/sparse/sparse_matrix.py (hunk 1 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py]: Now all APIs have implemented, please check.

File: python/dgl/sparse/sparse_matrix.py (hunk 4 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:508]: This is my example:
| | 0 | 1 | 2 | 3 | 4 | 5 |
| ---- | ---- | :---- | ----: | :----: | ---- | ---- |
0 | 0 |   |    |    |     |    
1 |    |    | 1 |    |  2  |    
2 |   |    |    | 3 |     |    
3 |   |    |    |    |     |  4
4 | 5 |    |    |    |     |    

if we select [0,1,4] rows:
| | 0 | 1 | 2 | 3 | 4 | 5 |
| ---- | ---- | :---- | ----: | :----: | ---- | ---- |
0 | 0 |   |    |    |     |    
1 |    |    | 1 |    |  2  |    
4 | 5 |    |    |    |     |    

so answer is: 
SparseMatrix(indices=tensor([[0, 1, 1, 2],[0, 2, 4, 0]]),
                     values=tensor([0, 1, 2, 5]),
                     shape=(3, 6), nnz=4)

File: python/dgl/sparse/sparse_matrix.py [hunk 5]
```diff
@@ -479,6 +479,62 @@ def is_diag(self):
...
         return self.c_sparse_matrix.is_diag()
 
+    def select(self, dim: int, ids: torch.Tensor or slice):
+        """Returns a sub-matrix selected by a list of ids or by range.
+
+        Parameters
+        ----------
+        dim : int
+            The dim to select from matrix, should be 0 or 1.
+        ids : tensor.Tensor or slice
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:489]: what's slice? a interger?

Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:489]: zhenkun suggest to use slice, a type of python used as (start, end, step). I think it is hard to be recognized whether users want to select two rows (x, y) or the range between them [x, y) if we use Tensor.

Comment (czkkkkkk) [reply python/dgl/sparse/sparse_matrix.py:489]: The range is not a Tensor. It is a [Python Slice](https://www.w3schools.com/python/ref_func_slice.asp) with three arguments: start, stop, and step. So I think it is better to use adopt two interfaces for the slicing, i.e., `index_select(dim, tensor)` and `range_select(dim, slice)`.

File: python/dgl/sparse/sparse_matrix.py (hunk 4 above)
Comment (czkkkkkk) [reply python/dgl/sparse/sparse_matrix.py:508]: Got you. The row ids are actually reordered.

File: python/dgl/sparse/sparse_matrix.py [hunk 6]
```diff
@@ -479,6 +479,128 @@ def is_diag(self):
         """Returns whether the sparse matrix is a diagonal matrix."""
         return self.c_sparse_matrix.is_diag()
 
+    def rowwise_select(self, ids: torch.Tensor):
+        """Returns a row-wise sparse matrix selected by ids.
+
+        Parameters
+        ----------
+        ids : tensor.Tensor
```
[the remark this answers is not in the document]
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py]: Add the "NOTE: ..." comments.

File: python/dgl/sparse/sparse_matrix.py (hunk 5 above)
Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:489]: Ah, I see, I was not aware that slice is a python type, then this API makes sense.

File: python/dgl/sparse/sparse_matrix.py [hunk 7]
```diff
@@ -479,6 +479,62 @@ def is_diag(self):
...
+            Slice indicates sub-matrix selected by range [start, end).
+
+        The function has not supported autograd.
+
+        Returns
+        -------
+        SparseMatrix
+            The sub-matrix only contains rows or columns selected.
+
+        Examples
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:501]: nit: add a blank line between examples.

File: python/dgl/sparse/sparse_matrix.py [hunk 8]
```diff
@@ -479,6 +479,62 @@ def is_diag(self):
...
+                     shape=(2, 6), nnz=3)
+        >>> A.select(1, slice(3, 6))
+        SparseMatrix(indices=tensor([[1, 2, 3],
+                                     [1, 0, 2]]),
+                     values=tensor([2, 3, 4]),
+                     shape=(5, 3), nnz=3)
+        """
+        if isinstance(ids, torch.Tensor):
+            return SparseMatrix(self.c_sparse_matrix.index_select(dim, ids))
+        else:
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:531]: if isinstance(ids, slice):
blabla
return

Raise error unsupported input type.

File: python/dgl/sparse/sparse_matrix.py [hunk 9]
```diff
@@ -479,6 +479,62 @@ def is_diag(self):
...
+                     values=tensor([1, 2, 3]),
+                     shape=(2, 6), nnz=3)
+        >>> A.select(1, slice(3, 6))
+        SparseMatrix(indices=tensor([[1, 2, 3],
+                                     [1, 0, 2]]),
+                     values=tensor([2, 3, 4]),
+                     shape=(5, 3), nnz=3)
+        """
+        if isinstance(ids, torch.Tensor):
+            return SparseMatrix(self.c_sparse_matrix.index_select(dim, ids))
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:530]: Does this c code exist at this moment? if not raise not implemented error.

Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:530]: All implemented, see: https://github.com/dmlc/dgl/pull/6208

Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:530]: For this PR, it should be not implemented. you want to merge this PR first, right?
And in https://github.com/dmlc/dgl/pull/6208, you can update it to the actual implementation.

File: python/dgl/sparse/sparse_matrix.py (hunk 8 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:531]: I do not find this error type except "TypeError", should I use it?

File: python/dgl/sparse/sparse_matrix.py (hunk 9 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:530]: OK, I got it.

File: python/dgl/sparse/sparse_matrix.py [hunk 10]
```diff
@@ -479,6 +479,74 @@ def is_diag(self):
...
+                     shape=(5, 3), nnz=3)
+        """
+        if isinstance(ids, torch.Tensor):
+            raise NotImplementedError
+        elif isinstance(ids, slice):
+            start = 0 if ids.start is None else ids.start
+            end = ids.stop
+            raise NotImplementedError
+        else:
+            raise TypeError
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:548]: Add more info about the TypeError, see example:
https://github.com/dmlc/dgl/blob/44a9faad7ab60d6bbcc08e094c0eaa86a0d73063/python/dgl/graphbolt/itemset.py#L67

Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py]: Fixed.

File: python/dgl/sparse/sparse_matrix.py (hunk 7 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:501]: Fixed, and added the case comments as other examples.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/sparse/sparse_matrix.py [hunk 11]
```diff
@@ -479,6 +479,72 @@ def is_diag(self):
...
         return self.c_sparse_matrix.is_diag()
 
+    def select(self, dim: int, ids: torch.Tensor or slice):
+        """Returns a sub-matrix selected by a list of ids or by range.
+
+        Parameters
+        ----------
+        dim : int
+            The dim to select from matrix, should be 0 or 1.
+        ids : tensor.Tensor or slice
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:489]:
```suggestion
        ids : Union[tensor.Tensor, slice]
```

File: python/dgl/sparse/sparse_matrix.py [hunk 12]
```diff
@@ -479,6 +479,72 @@ def is_diag(self):
         """Returns whether the sparse matrix is a diagonal matrix."""
         return self.c_sparse_matrix.is_diag()
 
+    def select(self, dim: int, ids: torch.Tensor or slice):
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:482]:
```suggestion
    def select(self, dim: int, ids: Union[tensor.Tensor, slice]):
```

File: python/dgl/sparse/sparse_matrix.py [hunk 13]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+        Parameters
+        ----------
+        dim : int
+            The dim to select from matrix, should be 0 or 1.
+        ids : Union[tensor.Tensor, slice]
+            Tensor indicates the row ids to select from matrix.
+            NOTE: The ids can include duplicate values.
+            Slice indicates sub-matrix selected by range [start, end).
+
+        The function has not supported autograd.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:494]: Do you plan to support autograd in the future?

File: python/dgl/sparse/sparse_matrix.py [hunk 14]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
         """Returns whether the sparse matrix is a diagonal matrix."""
         return self.c_sparse_matrix.is_diag()
 
+    def select(self, dim: int, ids: Union[tensor.Tensor, slice]):
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:482]:
```suggestion
    def select(self, dim: int, rule: Union[tensor.Tensor, slice]):
```

File: python/dgl/sparse/sparse_matrix.py [hunk 15]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+        """Returns a sub-matrix selected by a list of ids or by range.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:483]:
```suggestion
        """Returns a sub-matrix selected according to the given rule.
```

File: python/dgl/sparse/sparse_matrix.py [hunk 16]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+
+        Parameters
+        ----------
+        dim : int
+            The dim to select from matrix, should be 0 or 1.
+        ids : Union[tensor.Tensor, slice]
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:489]:
```suggestion
        rules : Union[tensor.Tensor, slice]
```

File: python/dgl/sparse/sparse_matrix.py (hunk 13 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py:494]: Sampling based GNN training not use slicing autograd yet, as suggested.

**Comment by xiangyuzhi** [comment]
@frozenbugs CI told me "Unused argument 'dim' (unused-argument)", I wonder how to pass it without changing API.

File: python/dgl/sparse/sparse_matrix.py [hunk 17]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+            Tensor indicates the row ids to select from matrix.
+            NOTE: The ids can include duplicate values.
+            Slice indicates sub-matrix selected by range [start, end).
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:492]:
```suggestion
            The selection rule.
            - if `rule` is a torch.Tensor, it indicates which IDs from the `dim` should be chosen from the matrix. Note that duplicated ids are allowed.
            - if `rule` is a slice object, the matrix will be viewed as a sequence along the given `dim`, and sliced according to the python slice definition.
```

Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:492]: Max char of a line is 80, please fix.

File: python/dgl/sparse/sparse_matrix.py [hunk 18]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+
+        The function has not supported autograd.
+
+        Returns
+        -------
+        SparseMatrix
+            The sub-matrix only contains rows or columns selected.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:499]:
```suggestion
            The sub-matrix which contains selected rows or columns.
```

File: python/dgl/sparse/sparse_matrix.py [hunk 19]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+
+        Examples
+        --------
+
+        Case1: Select several rows by index.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:504]:
```suggestion
        Case 1: Select rows by IDs.
```

File: python/dgl/sparse/sparse_matrix.py [hunk 20]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+        >>> val = torch.tensor([0, 1, 2, 3, 4, 5])
+        >>> A = dglsp.spmatrix(indices, val)
+        >>> row_ids = torch.tensor([0, 1, 4])
+        >>> A.select(0, row_ids)
+        SparseMatrix(indices=tensor([[0, 1, 1, 2],
+                                     [0, 2, 4, 0]]),
+                     values=tensor([0, 1, 2, 5]),
+                     shape=(3, 6), nnz=4)
+
+        Case2: Select several columns by index.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:516]:
```suggestion
        Case 2: Select columns by IDs.
```

File: python/dgl/sparse/sparse_matrix.py [hunk 21]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+
+        >>> column_ids = torch.tensor([0, 4, 5])
+        >>> A.select(1, column_ids)
+        SparseMatrix(indices=tensor([[0, 1, 3, 4],
+                                     [0, 1, 2, 0]]),
+                     values=tensor([0, 2, 4, 5]),
+                     shape=(5, 3), nnz=4)
+
+        Case3: Select several rows by range.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:525]:
```suggestion
        Case 3: Select rows with given slice object.
```

File: python/dgl/sparse/sparse_matrix.py [hunk 22]
```diff
@@ -479,6 +479,71 @@ def is_diag(self):
...
+
+        >>> A.select(0, slice(1, 3))
+        SparseMatrix(indices=tensor([[0, 0, 1],
+                                     [2, 4, 3]]),
+                     values=tensor([1, 2, 3]),
+                     shape=(2, 6), nnz=3)
+
+        Case4: Select several columns by range.
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:533]:
```suggestion
        Case 4: Select columns with given slice object.
```

File: python/dgl/sparse/sparse_matrix.py (hunk 19 above)
Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:504]: Move this line after L508 `A = dglsp.spmatrix(indices, val)`, the sp matrix initiation code is shared across all cases.

File: python/dgl/sparse/sparse_matrix.py (hunk 13 above)
Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:498]: The function has not supported autograd. -> The function does not support autograd.

`has not supported` sounds like you have a plan to do it soon.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/sparse/sparse_matrix.py (hunk 14 above)
Comment (xiangyuzhi) [reply python/dgl/sparse/sparse_matrix.py]: CI report error: "NameError: name 'Union' is not defined".

File: python/dgl/sparse/sparse_matrix.py [hunk 23]
```diff
@@ -479,6 +479,103 @@ def is_diag(self):
         """Returns whether the sparse matrix is a diagonal matrix."""
         return self.c_sparse_matrix.is_diag()
 
+    def index_select(self, dim: int, index: torch.Tensor):
+        """Returns a sub-matrix selected according to the given index.
+
+        Parameters
+        ----------
+        dim : int
+            The dim to select from matrix, should be 0 or 1.
```
Comment (czkkkkkk) [inline python/dgl/sparse/sparse_matrix.py:488]: Please specify dim=0/1 is rowwise or columnwise selection.

**Review by czkkkkkk — APPROVED** [review]
LGTM.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**